### PR TITLE
merge some of flake-utils into nixpkgs

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

90% of the flakes out there are using flake-utils' eachDefaultSystem.

Since all the flakes that use flake-utils also use nixpkgs as a dependency,
move that functionality here.

Before:

```nix
{
  description = "Flake utils demo";

  inputs.flake-utils.url = "github:numtide/flake-utils";

  outputs = { self, nixpkgs, flake-utils }:
    flake-utils.lib.eachDefaultSystem (system:
      let pkgs = nixpkgs.legacyPackages.${system}; in
      rec {
        packages = {
          hello = pkgs.hello;
        };
        defaultPackage = packages.hello;
      }
    );
}
```

After:

```nix
{
  description = "Flake utils demo";

  outputs = { self, nixpkgs }:
    nixpkgs.lib.mapAttrsJump (nixpkgs.lib.systems.tier1 ++ lib.systems.tier2) (system:
      let pkgs = nixpkgs.legacyPackages.${system}; in
      rec {
        packages = {
          hello = pkgs.hello;
        };
        defaultPackage = packages.hello;
      }
    );
}
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zimbatm** [comment]
Type of feedback I am looking for:

* mapAttrsJump: I invented this name because I didn't find prior code literature. If you find other examples out there I am interested.
* lib.systems.tier{1,2}: I'm not super familiar with the more complicated constructs. Are the strings acceptable here?

**Comment by zimbatm** [comment]
Potentially, this PR could also gain a `lib.systems.eachDefault` for the common case. And if you have a full list of all the tiers, happy to add those as well.

File: lib/attrsets.nix [hunk 1]
```diff
@@ -302,6 +302,35 @@ rec {
...
+   *
+   * This is mainly useful for flake-like structures.
+   *
+   * Type:
+   *   mapAttrsJump ::
+   *     [String] -> (String -> AttrSet) -> AttrSet<String, AttrSet>
+   *
+   * Example:
+   *   mapAttrsJump ["x86_64-linux"] (system: { hello = 42; })
+   *   => { x86_64-linux = { hello = 42; }; }
```
Comment (jtojnar) [inline lib/attrsets.nix:316]: This example looks like the regular `genAttrs`. Probably should be something like `{ hello = { x86_64-linux = 42; }; }`.

File: lib/attrsets.nix [hunk 2]
```diff
@@ -302,6 +302,35 @@ rec {
...
     in recurse [] set;
 
+  /* Builds a map from <attr>=value to <attr>.<item>=value for each item in
+   * the list.
+   *
+   * This is mainly useful for flake-like structures.
+   *
+   * Type:
+   *   mapAttrsJump ::
+   *     [String] -> (String -> AttrSet) -> AttrSet<String, AttrSet>
```
Comment (jtojnar) [inline lib/attrsets.nix:312]: Same with the type here

```suggestion
   *     [a] -> (a -> AttrSet<b, c>) -> AttrSet<b, AttrSet<a, c>>
```

Comment (jtojnar) [reply lib/attrsets.nix:312]: From the type, I would call it something like `genAttrsTransposed`.

Comment (zimbatm) [reply lib/attrsets.nix]: Good idea

**Comment by lilyball** [comment]
I would really like to have some of this in nixpkgs, but I don't think this is the right approach. `mapAttrsJump` is mildly confusing the way it's written, and it's rather specialized, and it requires picking the systems to support which is just likely to cause a lot of flakes to pick the wrong set (such as picking just tier1) when they don't actually have any reason to restrict it. I'd rather see this be a little more flake-focused, and in particular I'd prefer to see flake-utils's `eachSystem` and `eachDefaultSystem` just copied as-is.

`eachDefaultSystem` could potentially be renamed to something like `eachHydraSystem`, but that's a little more opaque and not as approachable by folks who aren't familiar with hydra. The name `eachDefaultSystem` is much more obviously the right thing to pick for most people. It could still be defined in terms of `lib.systems.supported.hydra` but this could be considered an implementation detail (e.g. nixpkgs should feel free to add systems to the list even if they're not supported by hydra, though I don't know if there's a reason to ever do so).

Given the flake-specific nature of this, I would propose that this go into a new `lib.flakes` library. And flake-utils's `allSystems` should probably be copied into `lib.systems` somewhere. I don't know if the list corresponds with anything that exists in nixpkgs already, but I don't see anything obvious in `lib.systems` that matches it.

---

Also as a side note, `lib.systems.supported.tier1`/`tier2`/`tier3` appears unusable. I see this PR has a TODO referencing an RFC, but the tier1/tier2 lists don't even include all hydra platforms, and in particular don't include aarch64-darwin. nixpkgs should most definitely not encourage a standard flake pattern that does not support the current crop of Apple hardware.

**Comment by zimbatm** [comment]
I'll close this because I don't have the energy to push through these changes. Hopefully, this will send the signal to somebody else to take over. I still think that something the useful subset of flake-utils should be added to nixpkgs directly.

**Pull request closed without merging**